api/src/main/java/org/apache/iceberg/catalog/SessionCatalog.java

```diff
@@ -0,0 +1,323 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.catalog;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.NamespaceNotEmptyException;
+import org.apache.iceberg.exceptions.NoSuchNamespaceException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+
+/**
+ * A Catalog API for table and namespace operations that includes session context.
+ */
+public interface SessionCatalog {
+  /**
+   * Context for a session.
+   */
+  final class SessionContext {
+    private final String sessionId;
+    private final String identity;
+    private final String credential;
```

Comment:
We might want to consider making credentials a map (this is pretty common and also how Trino represents credentials).

Reply:
This is really helpful. I'm not at all sure how we want to track credentials and identity, so more ideas here are great.

Reply:
The credential is also a part of identity, for reference: https://github.com/trinodb/trino/blob/master/core/trino-spi/src/main/java/io/trino/spi/security/Identity.java